lib/Sema/CSBindings.cpp

```diff
@@ -209,9 +209,45 @@ void ConstraintSystem::PotentialBindings::addPotentialBinding(
   if (auto *literalProtocol = binding.DefaultedProtocol)
     foundLiteralBinding(literalProtocol);
 
+  // If the type variable can't bind to an lvalue, make sure the
+  // type we pick isn't an lvalue.
+  if (!TypeVar->getImpl().canBindToLValue() &&
+      binding.BindingType->hasLValueType()) {
+    binding = binding.withType(binding.BindingType->getRValueType());
+  }
+
+  if (!isViable(binding))
+    return;
+
   Bindings.push_back(std::move(binding));
 }
 
+bool ConstraintSystem::PotentialBindings::isViable(
+    PotentialBinding &binding) const {
```

Comment:
What's special about bound generic types? What about nominal types nested inside generic types, etc?

Reply:
Just moved this code over from `tryTypeVariableBindings` but from what I've seen, it tries to detect bindings like `Set<$T1>, Set<$T2>, ...` and only attempt first `Set<$T1>`, I've looked at the constraint system in the improved example and it helps because the number of bindings is reduced which results in type variable not preferred as eagerly in certain cases.

Reply:
What if I have `Foo<$T1>.Bar`, `Foo<$2>.Bar`, etc. I'm just suggesting we can generalize this to look at any nominal type in generic context, not just bound generic type.

Reply:
Sure, makes sense, I can tackle that in the separate PR, example would be:

```swift
struct A<T> {
  struct B {
    init(_: T) {}
  }
}

let _ = [A.B(1), A.B(1.0)]
```